From 71c7cfac2b7ead9800275b5b040e972ed3669bf0 Mon Sep 17 00:00:00 2001
From: Matus Kral <matuskral@me.com>
Date: Wed, 30 Jul 2014 00:35:47 +0200
Subject: [PATCH] Resolution changing on play.

Control parameter in advancedsettings.xml:
  <video>
    <adjustdisplayrestomatchvideo>true</adjustdisplayrestomatchvideo>
  </video>

This is enabled by default on i.MX6.
---
 xbmc/cores/VideoRenderers/BaseRenderer.cpp | 30 ++++++++++++++++++++++++++++++
 xbmc/settings/AdvancedSettings.cpp         |  7 +++++++
 xbmc/settings/AdvancedSettings.h           |  2 ++
 3 files changed, 39 insertions(+)

diff --git a/xbmc/cores/VideoRenderers/BaseRenderer.cpp b/xbmc/cores/VideoRenderers/BaseRenderer.cpp
index 83c3adb..8e43060 100644
--- a/xbmc/cores/VideoRenderers/BaseRenderer.cpp
+++ b/xbmc/cores/VideoRenderers/BaseRenderer.cpp
@@ -34,6 +34,7 @@                                      
 #include "settings/AdvancedSettings.h"                
 #include "cores/VideoRenderers/RenderFlags.h"         
.                                                      
+#include "Application.h"
.                                                      
 CBaseRenderer::CBaseRenderer()                        
 {
@@ -221,12 +226,35 @@ void CBaseRenderer::FindResolutionFromFpsMatch(float fps, float& weight)
 
 RESOLUTION CBaseRenderer::FindClosestResolution(float fps, float multiplier, RESOLUTION current, float& weight)
 {
-  RESOLUTION_INFO curr = g_graphicsContext.GetResInfo(current);
+  RESOLUTION_INFO curr = g_application.m_res;
 
   float fRefreshRate = fps;
 
   float last_diff = fRefreshRate;
 
+  // CHANGERESOLUTION
+  if (g_advancedSettings.m_adjustResolutionVideoMatch)
+  {
+    for (size_t i = (int)RES_DESKTOP; i < CDisplaySettings::Get().ResolutionInfoSize(); i++)
+    {
+      const RESOLUTION_INFO info = g_graphicsContext.GetResInfo((RESOLUTION)i);
+
+      if (abs(info.iScreenWidth*info.iScreenHeight - m_sourceWidth*m_sourceHeight) > abs(curr.iScreenHeight*curr.iScreenWidth - m_sourceWidth*m_sourceHeight)
+      ||  info.iScreen != curr.iScreen
+      ||  (info.dwFlags & D3DPRESENTFLAG_MODEMASK) != (curr.dwFlags & D3DPRESENTFLAG_MODEMASK)
+      ||  fabs(info.fPixelRatio - curr.fPixelRatio) > 0.001)
+        {
+          CLog::Log(LOGDEBUG, "curr %.2f, trying %.2f, mode nr. %d, %dx%d msk %d, m_msk %d", info.fPixelRatio, curr.fPixelRatio, i,
+               info.iScreenWidth, info.iScreenHeight, info.dwFlags & D3DPRESENTFLAG_MODEMASK,
+                    m_iFlags & D3DPRESENTFLAG_MODEMASK);
+          continue;
+        }
+
+      current = (RESOLUTION)i;
+      curr    = info;
+    }
+  }
+
   // Find closest refresh rate
   for (size_t i = (int)RES_DESKTOP; i < CDisplaySettings::Get().ResolutionInfoSize(); i++)
   {
diff --git a/xbmc/settings/AdvancedSettings.cpp b/xbmc/settings/AdvancedSettings.cpp
index 478d725..9a424c3 100644
--- a/xbmc/settings/AdvancedSettings.cpp
+++ b/xbmc/settings/AdvancedSettings.cpp
@@ -129,6 +129,12 @@ void CAdvancedSettings::Initialize()
   m_audioDefaultPlayer = "paplayer";
   m_audioPlayCountMinimumPercent = 90.0f;
 
+#if defined(HAS_IMXVPU)
+  m_adjustResolutionVideoMatch = true;
+#else
+  m_adjustResolutionVideoMatch = false;
+#endif
+
   m_videoSubsDelayRange = 60;
   m_videoAudioDelayRange = 10;
   m_videoSmallStepBackSeconds = 7;
@@ -545,6 +551,7 @@ void CAdvancedSettings::ParseSettingsFile(const CStdString &file)
     XMLUtils::GetString(pElement, "defaultplayer", m_videoDefaultPlayer);
     XMLUtils::GetString(pElement, "defaultdvdplayer", m_videoDefaultDVDPlayer);
     XMLUtils::GetBoolean(pElement, "fullscreenonmoviestart", m_fullScreenOnMovieStart);
+    XMLUtils::GetBoolean(pElement, "adjustdisplayrestomatchvideo", m_adjustResolutionVideoMatch);
     // 101 on purpose - can be used to never automark as watched
     XMLUtils::GetFloat(pElement, "playcountminimumpercent", m_videoPlayCountMinimumPercent, 0.0f, 101.0f);
     XMLUtils::GetInt(pElement, "ignoresecondsatstart", m_videoIgnoreSecondsAtStart, 0, 900);
diff --git a/xbmc/settings/AdvancedSettings.h b/xbmc/settings/AdvancedSettings.h
index ca995e3..7106bff 100644
--- a/xbmc/settings/AdvancedSettings.h
+++ b/xbmc/settings/AdvancedSettings.h
@@ -146,6 +146,8 @@ class CAdvancedSettings : public ISettingCallback, public ISettingsHandler
     bool  m_omxHWAudioDecode;
     bool  m_omxDecodeStartWithValidFrame;
 
+    bool  m_adjustResolutionVideoMatch;
+
     float m_videoSubsDelayRange;
     float m_videoAudioDelayRange;
     int m_videoSmallStepBackSeconds;
-- 
2.2.1

